gax/src/main/java/com/google/api/gax/rpc/Watchdog.java

```diff
@@ -0,0 +1,316 @@
+  }
+
+  enum State {
+    /** Stream has been started, but doesn't have any outstanding requests. */
+    IDLE,
+    /** Stream is awaiting a response from upstream. */
+    WAITING,
+    /**
+     * Stream received a response from upstream, and is awaiting outerResponseObserver processing.
+     */
+    DELIVERING
+  }
+
+  class Stream implements ResponseObserver<ResponseT> {
+    private final Object lock = new Object();
+
+    private final Duration waitTimeout;
+    private boolean hasStarted;
+    private boolean autoAutoFlowControl = true;
+
+    private final ResponseObserver<ResponseT> outerResponseObserver;
+    private StreamController innerController;
+
+    @GuardedBy("lock")
+    private State state = State.IDLE;
+
+    @GuardedBy("lock")
+    private int pendingCount = 0;
+
+    @GuardedBy("lock")
+    private long lastActivityAt;
+
+    private volatile Throwable error;
+
+    Stream(ResponseObserver<ResponseT> responseObserver, Duration waitTimeout) {
+      this.waitTimeout = waitTimeout;
+      this.outerResponseObserver = responseObserver;
+      this.lastActivityAt = clock.millisTime();
+    }
+
+    @Override
+    public void onStart(StreamController controller) {
+      Preconditions.checkState(!hasStarted, "Already started");
+
+      this.innerController = controller;
+      outerResponseObserver.onStart(
+          new StreamController() {
+            @Override
+            public void disableAutoInboundFlowControl() {
+              Preconditions.checkState(
+                  !hasStarted, "Can't disable automatic flow control after the stream has started");
+              autoAutoFlowControl = false;
+              innerController.disableAutoInboundFlowControl();
+            }
+
+            @Override
+            public void request(int count) {
+              Stream.this.onRequest(count);
+            }
+
+            @Override
+            public void cancel() {
+              Stream.this.onCancel();
+            }
+          });
+
+      hasStarted = true;
+    }
+
+    private void onRequest(int count) {
+      Preconditions.checkArgument(count > 0, "count must be > 0");
+      Preconditions.checkState(!autoAutoFlowControl, "Auto flow control is enabled");
+
+      // Only reset the request water mark if there is no outstanding requests.
+      synchronized (lock) {
+        if (state == State.IDLE) {
+          state = State.WAITING;
+          lastActivityAt = clock.millisTime();
+        }
+
+        // Increment the request count without overflow
+        int maxIncrement = Integer.MAX_VALUE - pendingCount;
+        count = Math.min(maxIncrement, count);
+        pendingCount += count;
+      }
+      innerController.request(count);
+    }
+
+    private void onCancel() {
+      error = new CancellationException("User cancelled stream");
+      innerController.cancel();
+    }
+
+    @Override
+    public void onResponse(ResponseT response) {
+      synchronized (lock) {
+        state = State.DELIVERING;
```

Comment:
Could a response be received in the middle of stale cancellation?

Reply:
If I understand your concern correctly, no, the caller thread only sets the cancellation flag, the actual cancellation is plumbed through grpc by calling cancel() on the controller, so the actual cancellation will come from the same thread that would deliver the response and grpc has strict rules about callback order

Reply:
I added #453 that makes sure that the rules are enforced for custom callables